pkg/controllers/user/multiclusterapp/controller_multiclusterapp.go

```diff
@@ -0,0 +1,352 @@
+		projectLister:         cluster.Management.Management.Projects("").Controller().Lister(),
+		clusterName:           cluster.ClusterName,
+		templateVersionLister: cluster.Management.Management.TemplateVersions("").Controller().Lister(),
+	}
+	m.multiClusterApps.AddHandler(ctx, "multi-cluster-app-controller", m.sync)
+}
+
+func (m *MCAppController) sync(key string, mcapp *v3.MultiClusterApp) (runtime.Object, error) {
+	if mcapp == nil || mcapp.DeletionTimestamp != nil {
+		_, mcappName, err := cache.SplitMetaNamespaceKey(key)
+		if err != nil {
+			return mcapp, err
+		}
+		return m.deleteApps(mcappName, mcapp)
+	}
+	metaAccessor, err := meta.Accessor(mcapp)
+	if err != nil {
+		return mcapp, err
+	}
+	creatorID, ok := metaAccessor.GetAnnotations()[creatorIDAnn]
+	if !ok {
+		return mcapp, fmt.Errorf("MultiClusterApp %v has no creatorId annotation. Cannot create apps for %v", metaAccessor.GetName(), mcapp.Name)
+	}
+
+	answerMap := m.createAnswerMap(mcapp.Spec.Answers)
+
+	externalID, mcapp, err := m.getExternalID(mcapp)
+	if err != nil {
+		return mcapp, err
+	}
+
+	appName, mcapp, err := m.getAppName(mcapp, externalID)
+	if err != nil {
+		return mcapp, err
+	}
+
+	// for all targets, create the App{} instance, so that helm controller App lifecycle can pick it up
+	// only one app per project named mcapp-{{mcapp.Name}}
+	var mcappToUpdate *v3.MultiClusterApp
+	appNameInProject := multiClusterAppPrefix + appName
+	ann := make(map[string]string)
+	ann[creatorIDAnn] = creatorID
+	set := labels.Set(map[string]string{multiClusterAppIDSelector: mcapp.Name})
+	for ind, t := range mcapp.Spec.Targets {
+		split := strings.SplitN(t.ProjectName, ":", 2)
+		if len(split) != 2 {
+			return mcapp, fmt.Errorf("error in splitting project ID")
+		}
+		projectNS := split[1]
+		// check if the target project in this iteration is same as the cluster in current context, if not then don't create namespace and app else it
+		// will be in the wrong cluster
+		if split[0] != m.clusterName {
+			logrus.Debug("Not for the current cluster")
+			continue
+		}
+
+		// check if this app already exists
+		a, err := m.apps.GetNamespaced(projectNS, appNameInProject, metav1.GetOptions{})
+		if err != nil && !apierrors.IsNotFound(err) {
+			return mcapp, fmt.Errorf("error checking if app already exists in project: %v", err)
+		} else if a != nil && a.Name == appNameInProject {
+			logrus.Debugf("App for multiclusterapp %v already exists in project %v", mcapp.Name, t.ProjectName)
+			continue
+		}
+
+		// call createNsAndApp method
+		newTarget, mcapp, err := m.createNamespaceAndApp(&t, mcapp, answerMap, ann, set, projectNS, creatorID, appNameInProject, externalID)
+		if err != nil {
+			return mcapp, err
+		}
+		if newTarget != nil {
+			mcappToUpdate = mcapp.DeepCopy()
+			mcappToUpdate.Spec.Targets[ind].AppName = newTarget.AppName
+		}
+	}
+
+	if !reflect.DeepEqual(mcapp, mcappToUpdate) && mcappToUpdate != nil {
```

Comment:
We should swap the order of the two conditions - check mcappToUpdate != nil before calling reflect.DeepEqual which is more expensive call